Reusing logic with custom hooks (#466)
reactjs/fr.react.dev: src/content/learn/reusing-logic-with-custom-hooks.md

  }, []);

  function handleSaveClick() {
    console.log('✅ Progression enregistrée');
  }

  return (
    <button disabled={!isOnline} onClick={handleSaveClick}>
      {isOnline ? 'Enregistrer la progression' : 'Reconnexion...'}
    </button>
  );
}
```

</Sandpack>

> Vérifiez que le bouton changera d’apparence si vous éteignez le réseau.

Author: C'est vrai, *c'est pas Versailles ici* 😄

Vous devez respecter les conventions de nommage suivantes :

1. **Le nom des composants React doit commencer par une lettre en majuscule,** comme `StatusBar` et `SaveButton`. Les composants React doivent également renvoyer quelque chose que React sait afficher, comme un morceau de JSX.
2. **Le nom des Hook doit commencer par `use` suivi d’une majuscule,** comme [`useState`](/reference/react/useState) (intégré) ou `useOnlineStatus` (personnalisé, comme plus haut dans la page). Les Hooks peuvent renvoyer des valeurs arbitraires.

Cette convention garantit que vous pouvez toujours regarder un composant et savoir où son état, ses effets et d’autres fonctionnalités de React peuvent se « cacher ». Par exemple, si vous voyez un appel à la fonction `getColor()` dans votre composant, vous pouvez être sûr qu’il ne contient pas d’état React car son nom ne commence pas par `use`. Cependant, un appel de fonction comme `useOnlineStatus()` contiendra très probablement des appels à d’autres Hooks à l’intérieur.

<Note>

> Si votre linter est [configuré pour React,](/learn/editor-setup#linting) il appliquera cette convention de nommage. Remontez jusqu’au bac à sable et renommez `useOnlineStatus` en `getOnlineStatus`. Remarquez que le linter ne vous permettra plus appeler `useState` ou `useEffect` à l’intérieur. Seuls les Hooks et les composants peuvent appeler d’autres Hooks !

Author: ça tombe bien que tu me le dises (je verrai dans la suite de la revue). _Personnellement_, voir la ponctuation dans le style, ça m'embête un peu parfois (enfin, quand ce n'est pas toute la phrase qui est concernée).

Par exemple:

`Il ne faut **jamais désactiver votre linter.**` (qui ne me plaît pas _personnellement_) vs. `Il ne faut **jamais désactiver votre linter**.`, ce qui donne : 
- 👎 Il ne faut **jamais désactiver votre linter.**
- 👍 Il ne faut **jamais désactiver votre linter**.

Tu parles ici des liens, dois-je comprendre que ce sera le cas aussi pour le style ?

> Si votre linter est [configuré pour React,](/learn/editor-setup#linting) il appliquera cette convention de nommage. Remontez jusqu’au bac à sable et renommez `useOnlineStatus` en `getOnlineStatus`. Remarquez que le linter ne vous permettra plus appeler `useState` ou `useEffect` à l’intérieur. Seuls les Hooks et les composants peuvent appeler d’autres Hooks !

Reviewer: Yup!

      onReceiveMessage(msg);
    });
    return () => connection.disconnect();
  }, [roomId, serverUrl, onReceiveMessage]); // ✅ Toutes les dépendances sont déclarées.
}
```

Ça fonctionnera, mais il y a une autre amélioration que vous pouvez apporter quand votre Hook personnalisé accepte des gestionnaires d’événements.

> Ajouter une dépendance à `onReceiveMessage` n’est pas idéal car il entraînera une reconnexion au salon à chaque réaffichage du composant. [Enrobez ce gestionnaire d’état dans un événement d’effet pour le supprimer des dépendances :](/learn/removing-effect-dependencies#wrapping-an-event-handler-from-the-props)

Author: Je vois que tu as mis `Événement d’Effet` avec des majuscules... Je n'étais pas trop sûr (même si j'écris **H**ook !). Il va falloir que je change dans l'autre PR, où ça doit apparaître 2000 fois 😅

}
```

C’est une raison pour laquelle il est souvent utile d’enrober des effets dans des Hooks personnalisés :

1. Vous rendez le flux de données vers et depuis vos effets très explicite.
2. Vous permettez à vos composants de se concentrer sur l’intention plutôt que sur l’implémentation exacte de vos effets.
3. Lorsque React ajoute de nouvelles fonctionnalités, vous pouvez retirer ces effets sans changer aucun de vos composants.

À la manière d’un [système de design,](https://uxdesign.cc/everything-you-need-to-know-about-design-systems-54b109851969) vous pourriez trouver utile de commencer à extraire les idiomes communs des composants de votre app dans des Hooks personnalisés. Ainsi, le code de vos composants restera centré sur l’intention et vous éviterez la plupart du temps d’utiliser des effets bruts. De nombreux Hooks personnalisés de qualité sont maintenus par la communauté de React.

<DeepDive>

#### React fournira-t-il une solution intégrée pour la récupération des données ? {/*will-react-provide-any-built-in-solution-for-data-fetching*/}

> Nous travaillons encore sur les détails, mais nous pensons qu’à l’avenir, vous écrirez la récupération de données de cette façon :

Reviewer: ```suggestion
Nous travaillons encore sur les détails, mais nous pensons qu’à l’avenir, vous pourrez charger des données comme ceci :
```

  // ...
```

Si vous utilisez des Hooks personnalisés comme le `useData` plus haut dans votre app, la migration vers l’approche finalement recommandée nécessitera moins de changements que si vous écrivez manuellement des effets bruts dans chaque composant. Cependant, l’ancienne approche continuera de bien fonctionner, donc si vous vous sentez à l’aise en écrivant des effets bruts, vous pouvez continuer ainsi.

</DeepDive>

### Il y a plus d’une façon de faire {/*there-is-more-than-one-way-to-do-it*/}

> Supposons que vous voulez implémenter une animation de fondu-enchaîné *en partant de zéro* en utilisant l’API du navigateur [`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/window/requestAnimationFrame). Vous pouvez commencer par un effet qui initialise une boucle d’animation. Pendant chaque image de l’animation, vous pourriez changer l’opacité du nœud du DOM si vous le [conservez dans une ref](/learn/manipulating-the-dom-with-refs) jusqu’à ce qu’elle atteigne `1`. Votre code pourrait commencer ainsi :

Reviewer: ```suggestion
Supposons que vous souhaitiez implémenter une animation de fondu enchaîné *en partant de zéro* avec l’API [`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/window/requestAnimationFrame) du navigateur. Vous pouvez commencer par un Effet qui initialise une boucle d’animation. À chaque étape de l’animation, vous pourriez changer l’opacité du nœud du DOM que vous aurez [conservé dans une ref](/learn/manipulating-the-dom-with-refs) jusqu’à ce qu’elle atteigne `1`. Votre code pourrait commencer ainsi :
```

</Sandpack>

> Pour rendre le composant plus lisible, vous pouvez extraire la logique dans un Hook personnalisé `useFadeIn` :

Reviewer: ```suggestion
Pour rendre le composant plus lisible, vous pourriez extraire la logique dans un Hook personnalisé `useFadeIn` :
```

"might" est nettement plus incertain que "would" ou "could".

      const progress = Math.min(timePassed / duration, 1);
      onProgress(progress);
      if (progress < 1) {

>         // Nous avons encore des frames à dessiner.

Reviewer: ```suggestion
        // Nous avons encore des étapes à dessiner.
```

    if (progress === 1) {
      this.stop();
    } else {

>       // Nous avons encore des frames à dessiner.

Reviewer: ```suggestion
      // Nous avons encore des étapes à dessiner.
```

</Sandpack>

Les effets permettent à React de se connecter à des systèmes externes. Plus la coordination entre les effets est nécessaire (par exemple pour enchaîner des animations multiples), plus il est sensé d’extraire *complètement* cette logique des effets et des Hooks, comme dans le bac à sable ci-dessus. Le code extrait *devient* ainsi le « système externe ». Ça permet à vos effets de rester simples car ils n’auront qu’à envoyer des messages au système que vous avez sorti de React.

> Les exemples ci-dessus supposent que la logique de fondu-enchaîné soit écrite en JavaScript. Cependant, cette animation particulière de fondu-enchaîné est à la fois plus simple et beaucoup plus efficace lorsqu’elle est écrite par une simple [animation CSS :](https://developer.mozilla.org/fr/docs/Web/CSS/CSS_Animations/Using_CSS_animations)

Reviewer: ```suggestion
Les exemples ci-dessus supposent que la logique de fondu enchaîné soit écrite en JavaScript. Cependant, cette animation particulière est à la fois plus simple et beaucoup plus efficace lorsqu’elle est implémentée comme une simple [animation CSS](https://developer.mozilla.org/fr/docs/Web/CSS/CSS_Animations/Using_CSS_animations) :
```

</Sandpack>

Parfois, vous n’avez même pas besoin d’un Hook !

<Recap>

- Les Hooks personnalisés vous permettent de partager la logique entre les composants.
- Le nom des Hooks personnalisés doit commencer par `use` et être suivi d’une lettre majuscule.
- Les Hooks personnalisés ne partagent que la logique d’état et non l’état lui-même.
- Vous pouvez passer des valeurs réactives d’un Hook à un autre, et elles restent à jour.
- Tous les Hooks sont réexécutés à chaque réaffichage de votre composant.
- Le code de vos Hooks personnalisés doit être pur, comme le code de votre composant.

> - Enrobez les gestionnaires d’événements reçus par les Hooks personnalisés dans des événéments d’effet.

Reviewer: ```suggestion
- Enrobez les gestionnaires d’événements reçus par les Hooks personnalisés dans des Événéments d’Effet.
```

</Sandpack>

Remarquez que `App.js` n’a plus besoin d’importer `useState` ni `useEffect` désormais.

</Solution>

#### Rendez le délai du compteur configurable {/*make-the-counter-delay-configurable*/}

> Dans cet exemple, il y a une variable d’état `delay` qui est contrôlée par un curseur, mais sa valeur n’est pas utilisée. Passez la valeur de `delay` à votre Hook personnalisé, et changez le Hook `useCounter` pour utiliser le `delay` passé plutôt que les `1000` ms codés en dur.

Reviewer: ```suggestion
Dans cet exemple, il y a une variable d’état `delay` qui est contrôlée par un curseur, mais sa valeur n’est pas utilisée. Passez la valeur de `delay` à votre Hook personnalisé, et changez le Hook `useCounter` pour utiliser le `delay` passé plutôt que les `1000` ms codées en dur.
```

</Sandpack>

Notez que cette solution pose un petit problème que vous résoudrez au prochain défi.

</Solution>

#### Définissez un intervalle de réinitialisation {/*fix-a-resetting-interval*/}

Dans cet exemple, il y a *deux* intervalles séparés.

Le composant `App` appelle `useCounter`, qui appelle `useInterval` pour mettre à jour le compteur à chaque seconde. Mais le composant `App` appelle *aussi* `useInterval` pour mettre à jour aléatoirement la couleur de l’arrière-plan de la page toutes les deux secondes.

Pour certaines raisons, la fonction de rappel qui met à jour l’arrière-plan la page ne s’exécute jamais. Ajoutez quelques journaux dans `useInterval` :

```js {2,5}
  useEffect(() => {
    console.log('✅ Définition d’un intervalle avec un délai de ', delay)
    const id = setInterval(onTick, delay);
    return () => {
      console.log('❌ Effacement d’un intervalle avec un délai de ', delay)
      clearInterval(id);
    };
  }, [onTick, delay]);
```

> Les journaux correspondent-ils à ce que vous attendez ? Si certains effets semblent se resynchroniser inutilement, pouvez-vous imaginer quelle dépendance est à l’origine de ce comportement ? Existe-t-il un moyen de [supprimer cette dépendance](/learn/removing-effect-dependencies) de votre effet ?

Reviewer: ```suggestion
Les messages correspondent-ils à vos attentes ? Si certains Effets semblent se resynchroniser inutilement, pouvez-vous deviner quelle dépendance est à l’origine de ce comportement ? Existe-t-il un moyen de [supprimer cette dépendance](/learn/removing-effect-dependencies) de votre Effet ?
```

<Solution>

> À l’intérieur de `useInterval`, enrobez la fonction de rappel du tick dans un événement d’effet, comme vous l’avez fait [plus tôt dans cette page.](/learn/reusing-logic-with-custom-hooks#passing-event-handlers-to-custom-hooks)

Reviewer: ```suggestion
À l’intérieur de `useInterval`, enrobez la fonction de rappel du tick dans un Événement d’Effet, comme vous l’avez fait [plus tôt dans cette page](/learn/reusing-logic-with-custom-hooks#passing-event-handlers-to-custom-hooks).
```

Comme pour Effect, Effect Event a un sens très particulier, et la capitalisation doit être préservée.  Je crois que d'autres pages que tu vises en parlent aussi.

</Solution>

#### Implémentez un mouvement d’échelonnement {/*implement-a-staggering-movement*/}

Dans cet exemple, le Hook `usePointerPosition()` suit la position du curseur. Essayez de déplacer votre curseur ou votre doigt dans la zone de prévisualisation et voyez le point rouge suivre votre mouvement. Sa position est enregistrée dans la variable `pos1`.

> En réalité, il y a cinq (!) points rouges différents qui sont dessinés. Vous ne les voyez pas actuellement car ils apparaissent tous sur la même position. C’est ça que vous devez corriger. Ce que vous voulez implémenter à la place d’un mouvement « décalé » : chaque point doit « suivre » la trajectoire du point précédent. Par exemple, si vous déplacez rapidement votre curseur, le premier point doit le suivre immédiatement, le deuxième point doit suivre le premier point avec un léger décalage, le troisième point doit suivre le deuxième point, et ainsi de suite.

Reviewer: ```suggestion
En réalité, il y a cinq (!) points rouges différents qui sont dessinés. Vous ne les voyez pas pour le moment car ils apparaissent tous à la même position. C’est ça que vous devez corriger. Vous devez implémenter au lieu de ça un mouvement « échelonné » : chaque point doit « suivre » la trajectoire du point précédent. Par exemple, si vous déplacez rapidement votre curseur, le premier point doit le suivre immédiatement, le deuxième point doit suivre le premier point avec un léger décalage, le troisième point doit suivre le deuxième point, et ainsi de suite.
```

Vous devez implémenter le Hook personnalisé `useDelayedValue`. Son implémentation actuelle retourne la `value` qui lui a été donnée. À la place, vous voulez retourner la valeur qu’elle valait `delay` millisecondes plus tôt. Vous aurez peut-être besoin d’un état et d’un effet pour ça.

Une fois que vous aurez implémenté `useDelayedValue`, vous devriez voir les points se déplacer les uns après les autres.

<Hint>

> Vous aurez besoin de stocker `delayedValue` comme variable d’état à l’intérieur de votre Hook. Quand la `value` change, vous devrez exécuter un effet. Cet effet devra mettre à jour `delayedValue` après le `delay`. Vous pouvez trouver utile d’appeler `setTimeout`.

Reviewer: ```suggestion
Vous aurez besoin de stocker `delayedValue` comme variable d’état à l’intérieur de votre Hook. Quand la `value` change, vous devrez exécuter un Effet. Cet Effet devra mettre à jour `delayedValue` après le `delay`. Recourir à `setTimeout` pourrait s’avérer utile.
```

> Est-ce que cet effet a besoin de nettoyage ? Pourquoi ou pourquoi pas ?

Reviewer: ```suggestion
Est-ce que cet effet a besoin de nettoyage ? Pourquoi (ou pourquoi non) ?
```
